### PR TITLE
Reduce allocations

### PR DESCRIPTION
**Pull request by Pliner**
Hi!

This PR tries to reduce allocation inside GroboContainer in the following case:
```
using System;
using System.Reflection;

using GroboContainer.Core;
using GroboContainer.Impl;

namespace GroboContainer.Memory
{
    public static class EntryPoint
    {
        public static void Main()
        {
            IContainer container = new Container(new ContainerConfiguration(new []{Assembly.GetEntryAssembly()}, mode:ContainerMode.UseShortLog));

            container.Create<IContext, IRoot<string>>(null);
            Console.WriteLine("Press any key");
            Console.ReadKey();
            for (var i = 0; i < 1024; ++i)
            {
                container.Create<IContext, IRoot<string>>(null);
            }
            Console.WriteLine("Press any key");
            Console.ReadKey();
        }
    }

    public interface IContext
    {
    }

    public interface IRoot<T>
    {   
    }

    public class Root : IRoot<string>
    {
        public Root(IContext context, Func<IContext, IChild> createChild) => child = createChild(context);

        private readonly IChild child;
    }

    public interface IChild
    {
    }

    public class Child : IChild
    {
        public Child(IContext context, Func<IContext, IGrandchild> createChild) => child = createChild(context);

        private readonly IGrandchild child;
    }

    public interface IGrandchild
    {
    }

    public class Grandchild : IGrandchild
    {
        public Grandchild(IContext context)
        {
        }
    }
}
```
We use IContext as a way to pass a current user id and a current organization id.
Before PR:
before
[Before.zip](https://github.com/skbkontur/GroboContainer/files/2497842/Before.zip)

After PR:
after
[After.zip](https://github.com/skbkontur/GroboContainer/files/2497843/After.zip)

The following allocations were fixed:
1. `LogItem` was converted struct. There is no reason to have it as class.
2. `TypeArray` was removed. Imho, the only reason to have it was caching of hashCode, but it's not relevant anymore, because `ConcurrentDictionary` calculates hashCode per key only once and stores it.
3. Allocations in `InjectionContext` constructor. The constructor assigned  `CreateHolder` method to `createHolder` field. The assigment caused allocation of `Func<IInjectionContext, IContextHolder>`. 
4. Allocations of empty arrays. `EmptyArray` class was added(because of net45 target `Array<T>.Empty` cannot be used).
5. Allocations of `Type[]`. `TypeArray` was added.

### CODE REVIEW

**Comment by Pliner** [comment]
The PR fixes only simple cases of allocations. There are several major unresolved cases:
1. Cycle checking inside `InjectionContext`. The implementation works well with singleton instances, because it checks every singleton only once, but it case of per request instances it checks them every time.
2. `GroboContainerLog` inside `InjectionContext`. The situation merely the same as in (1), because the log is usable only when you can't construct the instance and you want to understand what is wrong. But after a first successfull creation of an instance, imho it's not so important to collect the log.

So I propose to allocate the log and to run `cycle checking` only during the first time allocation of the type.

File: GroboContainer/Impl/Abstractions/AutoConfiguration/AutoImplementationConfiguration.cs [hunk 1]
```diff
@@ -28,8 +28,8 @@ public object GetOrCreateInstance(IInjectionContext context, ICreationContext cr
                 lock (configurationLock)
                     if (instance == null)
                     {
-                        IClassFactory noParametersFactory = implementation.GetFactory(Type.EmptyTypes, creationContext);
-                        return instance = noParametersFactory.Create(context, new object[0]);
```
Comment (spaceorc) [inline GroboContainer/Impl/Abstractions/AutoConfiguration/AutoImplementationConfiguration.cs]: Type.EmptyTypes - он правда медленнее?

File: GroboContainer/Impl/Contexts/NoContextHolder.cs [hunk 2]
```diff
@@ -4,6 +4,8 @@ namespace GroboContainer.Impl.Contexts
 {
     public class NoContextHolder : IContextHolder
     {
+        public static readonly NoContextHolder Instance = new NoContextHolder();
```
Comment (spaceorc) [inline GroboContainer/Impl/Contexts/NoContextHolder.cs]: Давай на всякий случай вообще конструктор приватным тогда сделаем

File: GroboContainer/New/Implementation.cs [hunk 3]
```diff
@@ -58,5 +58,34 @@ public IClassFactory GetFactory(Type[] parameterTypes, ICreationContext creation
         }
 
         #endregion
+
+        internal sealed class TypeArrayEqualityComparer : IEqualityComparer<Type[]>
```
Comment (spaceorc) [inline GroboContainer/New/Implementation.cs]: А чем тебе стандартный, встроенный StructuralEqualityComparer не угодил? Он вроде должен лучше работать. Я конечно не проверял, но он же встроенный...
https://docs.microsoft.com/en-us/dotnet/api/system.collections.structuralcomparisons.structuralequalitycomparer?view=netframework-4.7.2

File: GroboContainer/Impl/Abstractions/AutoConfiguration/AutoImplementationConfiguration.cs (hunk 1 above)
Comment (Pliner) [reply GroboContainer/Impl/Abstractions/AutoConfiguration/AutoImplementationConfiguration.cs]: Он не медленнее, он реализован вот буквально так же, как `EmptyArray<T>.Instance`. Я для единообразия решил заменить(так как для `object[]` не откуда было взять этот пустой массив, поэтому, и пришлось сделать `EmptyArray`),но раз это вызвало вопросы - откачу обратно на `Type.EmptyTypes`.

**Pull request merged**